_nx_parallel/config.py

```diff
@@ -0,0 +1,57 @@
+from networkx.utils.configs import Config
+from typing import Union
+from dataclasses import dataclass, field, asdict
+from joblib import parallel_config
+import networkx as nx
+from functools import wraps
+import os
+
+__all__ = [
+    "_set_nx_config",
+    "_config",
+]
+
+
+@dataclass
+class ParallelConfig(Config):
+    nx_config: bool = False
+    backend: str = "loky"
+    n_jobs: int = None
+    verbose: int = 0
+    temp_folder: str = None
+    max_nbytes: Union[int, str] = "1M"
+    mmap_mode: str = "r"
+    prefer: str = None
+    require: str = None
+    inner_max_num_threads: int = None
+    backend_params: dict = field(default_factory=dict)
+
+
+_config = ParallelConfig()
+
+
+def _set_nx_config():
+    """Decorator to set the configuration for the parallel computation
+    of the nx-parallel algorithms."""
```

Comment:
I think I see why you put the `_set_nx_config` function here (the module is config.py afterall). But the decorator isn't used in the `_nx_parallel` package. It is used in `nx_parallel`. It feels like this function might be better found if it is in the main package (perhaps in utils?).

Also, the name sounds like we are setting the nx_config values. What do you think of `_allow_nx_config` or `_add_nx_config`?

Reply:
I initially did have the `_set_nx_config` decorator in `utils` but then I was getting the following circular import error. So, then I moved it to `_nx_parallel`.

```
>>> import networkx as nx
>>> G = nx.path_graph(10)
>>> nx.square_clustering(G, backend="parallel")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aditi/Desktop/nx-parallel/venv/lib/python3.12/site-packages/networkx/utils/decorators.py", line 789, in func
    return argmap._lazy_compile(__wrapper)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<class 'networkx.utils.decorators.argmap'> compilation 20", line 3, in argmap_square_clustering_17
  File "/Users/aditi/Desktop/nx-parallel/venv/lib/python3.12/site-packages/networkx/utils/backends.py", line 910, in __call__
    return self._convert_and_call(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aditi/Desktop/nx-parallel/venv/lib/python3.12/site-packages/networkx/utils/backends.py", line 1305, in _convert_and_call
    backend = _load_backend(backend_name)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aditi/Desktop/nx-parallel/venv/lib/python3.12/site-packages/networkx/utils/backends.py", line 445, in _load_backend
    rv = _loaded_backends[backend_name] = backends[backend_name].load()
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/metadata/__init__.py", line 205, in load
    module = import_module(match.group('module'))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/Users/aditi/Desktop/nx-parallel/nx_parallel/__init__.py", line 1, in <module>
    from .algorithms import *
  File "/Users/aditi/Desktop/nx-parallel/nx_parallel/algorithms/__init__.py", line 13, in <module>
    from .cluster import *
  File "/Users/aditi/Desktop/nx-parallel/nx_parallel/algorithms/cluster.py", line 10, in <module>
    @nxp._set_nx_config()
     ^^^^^^^^^^^^^^^^^^
AttributeError: partially initialized module 'nx_parallel' has no attribute '_set_nx_config' (most likely due to a circular import)
>>> 
```

And yes the name does seem a bit off. Thank you for pointing it out. 
Also, I'm not in fair of using `allow_` or `add_` because the decorator first checks the value of the `nx_config` and then accordingly adds or not adds the networkx's config.

Reply:
That makes sense...  The decorator has to be loaded into the namespace before it can be used. I think changing the order of imports inside `nx_parallel.__init__.py` to put `utils` first will fix the "partially initialized" problem (unless something inside `utils` uses the decorator). 

I think `allow_` indicates that the nx config won't always be used, but it can be used and without the decorator it can't be used. The term `set_` sounds like something is getting set (either nx_config values are getting set, or you are setting up the function to always use nx_config).  Maybe there is a better term?

Reply:
Moving `utils` on top is giving the same error. 
I think (and I'm probably wrong) that it might have to do with this lazy compiling-- `argmap._lazy_compile(__wrapper)(*args, **kwargs)`. But, idk much about how the `argmap._lazy_compile` works.

> ... Maybe there is a better term?

does `_configure` or `_configure_context` sounds good to you?

Reply:
Ahh...  try calling it with `@nxp.utils._set_config(`. Referring to the submodule gives it enough closer to allow the lookup to find _set_config. (it needs both the order change and the explicit module namespace.)

I like the `configure` word... but that decorator is not configuring. How about `_setup_dual_config`? hmm... maybe too long... but it's not *that* long...

Reply:
It worked but I don't understand exactly how and why... I'd like to look into it; could pls direct me to where I should look to understand this?

Also `dual` gives a sense that 2 config systems are being setup, but it only sets up one. What do you think of `_apply_config` or `apply_parallel_config`?

Reply:
Interesting  -- I thought it was setting up both parallel_config and nx_config? And then both are ready to be used...  

For where to look for info about importing namespaces, I don't know a place that talks about the mechanics of partially completed imports. `importlib` might be a place to look, but the docs seem to be more about how to use it than what goes on under the hood. I do know that importing a package runs it's `__init__.py` file. And that often imports other subpackages with their `__init__.py` files. Names can be added explicitly, or by using `*` in which case the subpackage's attribute `__all__` is used to figure out which names to add to the outer namespace.  But -- this issue involves code run by `__init__.py` that calls another part of nxp. First of all, that other part of nxp must have already been defined (that is, `utils` must have been imported before you can use it). What we ran into today is that it seems that the current/outer namespace is not fully accessible -- perhaps each layer only becomes accessible when that `__init__.py` file has finished, but I didn't know that restriction, so I am guessing as to why we can't use `nxp._set_config` as soon as `from utils import *` is finished. At the terminal I would expect to be able to use a function that was imported on one line as soon as that line completes. ???!?!  Anyway, by calling `nxp.utils._set_config` we only need to access the `utils` namespace -- instead of the main nxp namespace (which has not finished its initialization). The `utils` has finished its initialization. At least that is what I think is the reason we can access `nxp.utils._set_config`. 

One good way to reverse engineer the import system is to set up an example package with lots of print statements to see what gets run in which order. But I haven't done that yet... :)

Reply:
> Interesting -- I thought it was setting up both parallel_config and nx_config? And then both are ready to be used...

I renamed the `nx_config` configuration in `nx.config.backends.parallel` to `active`. I hope that make it more clear.

When `active=True` we use configs in `nx.config.backends.parallel`(i.e. put these configs in a `joblib.parallel_config` and make the function call in this context) and when it's `False` we don't do anything(just simply calls the function); the `joblib.Parallel` in nx-parallel algorithms itself looks for the configs set by the user in the `joblib.parallel_config`.

Reply:
So, the user can use `parallel_config` at any time in any function. But this decorator makes it also possible to use nx_config (and either/both config systems). That's why I am thinking of it as enabling dual configs systems (nx_config and parallel_config). We can't turn off the joblib config system. So we are adding a new one, resulting in two config systems.

Reply:
yes, we can't turn off the joblib config system, but we can overwrite it, and that is what is happening here. If someone enables networkx's config and then run the function in `joblib.parallel_config` like this:

```python
nx.config.backends.parallel.active = True
nx.config.backends.parallel.n_jobs = 6

with joblib.parallel_config(n_jobs=4):
    nx.square_clustering(G, backend="parallel")
```

so here the `n_jobs=4` is overwritten by the internal `with joblib.parallel_config(n_jobs=6)`(just for the nx-parallel algorithm). This is what the `_set_nx_config` decorator is doing.

Reply:
Does `_set_nx_config` overwrite all the parameter settings of `parallel_config`? or only the values explcitily set in `nx_config`? For example:
```python
nx.config.backends.parallel.active = True
nx.config.backends.parallel.n_jobs = 6

with joblib.parallel_config(n_jobs=4, verbose=55):
    nx.square_clustering(G, backend="parallel")
```
Does this get run with `verbose=55`?

Reply:
> Does this get run with verbose=55?

no, all of them get overwritten bcoz we pass all of the configs in `nx.config.backends.parallel` in the internal `parallel_config` run by the decorator(only when `active=True`). But, this is just for the nx-parallel functions, like here:

```py
with joblib.parallel_config(n_jobs=4, verbose=55):
    nx.square_clustering(G, backend="parallel") # n_jobs=6, verbose=0
    joblib.Parallel(joblib.delayed(sqrt)(i ** 2) for i in range(10)) # n_jobs = 4, verbose=55
```

Reply:
Hmmm...  that seems confusing to me...

What's the advantage of making nx_config "active" over just having the user use `joblib.parallel_config`?
I guess they could specify different config parameters for the networkx parts and for any non-networkx code they are running... but that doesn't seem very persuasive. They have to work around that for the rest of their code. Why make nx-parallel different from a config perspective?

Reply:
This does make sense...  Perhaps you can put this description (with an example) into the docs -- and explain that people can use joblib's config directly, or if they prefer the nx_config interface they can use that without affecting their other joblib code (because nx-parallel uses a context manager with the nx_config options).

Reply:
Yes -- I see what you are saying now -- some users will be used to storing their backend configuration parameters using the nx_config system. They will find it natural to use the nx_config system to provide the config parameters to joblib through nx-parallel. Other users may be more comfortable using the joblib parallel_config system. By providing this option we support either method of configuration.
(written last night but I didn't press "comment" :)